File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.List;
import java.util.ArrayList;
class Solution {
    public int solution(int cacheSize, String[] cities) {
        // 캐시 크기가 0이라면, 전부 cache miss
        if (cacheSize == 0) {
            return cities.length * 5;
        }
        
        List<String> cachedCities = new ArrayList<>();
        int answer = 0;
        for (String city : cities) {
            city = city.toLowerCase();
            if (!cachedCities.contains(city)) {
                if (cachedCities.size() == cacheSize) {
                    cachedCities.remove(0);   
                }                
                cachedCities.add(city);
                answer += 5;
            } else {
                cachedCities.remove(city);
                cachedCities.add(city);
                answer += 1;
            }
            
        }
        return answer;
    }
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.js
function solution(cacheSize, cities) {
    
    var answer = 0;
    
    // LRU 
    // 캐시 배열
    const cache = []
    
    for (c of cities) {
        const city = c.toUpperCase()
        const cache_idx = cache.indexOf(city)
        if (cache_idx === -1) {
            // cache miss
            answer += 5;
            cache.push(city)
            if (cache.length > cacheSize) {
                cache.shift()
            }            
        } else {
            // cache hit
            answer += 1;
            cache.push(cache.splice(cache_idx, 1)[0])            
        }          
    }
    
    
    return answer;
}